Mathlib/LinearAlgebra/Matrix/SchurTriangulation.lean

```diff
@@ -0,0 +1,300 @@
+  `LinearMap.SchurTriangulationAux.of` contains the main algorithm for the triangulation procedure.
+
+-/
+
+section
+
+variable {m n : ℕ}
+
+/-- `subNat' i h` subtracts `m` from `i`. This is an alternative form of `Fin.subNat`. -/
+@[inline] def Fin.subNat' (i : Fin (m + n)) (h : ¬ i < m) : Fin n :=
+  subNat m (Fin.cast (m.add_comm n) i) (Nat.ge_of_not_lt h)
+
+namespace Equiv
+
+/-- The composition of `finSumFinEquiv` and `Equiv.sumEquivSigmalCond` used by
+`LinearMap.SchurTriangulationAux.of`. -/
+def finAddEquivSigmaCond : Fin (m + n) ≃ Σ b, cond b (Fin m) (Fin n) :=
+  finSumFinEquiv.symm.trans (sumEquivSigmalCond (Fin m) (Fin n))
+
+variable {i : Fin (m + n)}
+
+lemma finAddEquivSigmaCond_true (h : i < m) : finAddEquivSigmaCond i = ⟨true, i, h⟩ :=
+  congrArg (sumEquivSigmalCond (Fin m) (Fin n)) (finSumFinEquiv_symm_apply_castAdd ⟨i, h⟩)
+
+lemma finAddEquivSigmaCond_false (h : ¬ i < m) : finAddEquivSigmaCond i = ⟨false, i.subNat' h⟩ :=
+  let j : Fin n := i.subNat' h
+  calc finAddEquivSigmaCond i
+    _ = finAddEquivSigmaCond (Fin.natAdd m j) :=
+      suffices m + (i - m) = i from congrArg _ (Fin.ext this.symm)
+      Nat.add_sub_of_le (Nat.le_of_not_gt h)
+    _ = ⟨false, i.subNat' h⟩ :=
+      congrArg (sumEquivSigmalCond (Fin m) (Fin n)) (finSumFinEquiv_symm_apply_natAdd j)
+
+end Equiv
+
+end
+
+variable {m n : Type _}
+
+/-- The type family parameterized by `Bool` is finite if each type variant is finite. -/
+instance [M : Fintype m] [N : Fintype n] (b : Bool) : Fintype (cond b m n) := b.rec N M
+
+/-- The type family parameterized by `Bool` has decidable equality if each type variant is
+decidable. -/
+instance [DecidableEq m] [DecidableEq n] : DecidableEq (Σ b, cond b m n)
+  | ⟨true, _⟩, ⟨false, _⟩
+  | ⟨false, _⟩, ⟨true, _⟩ => isFalse nofun
+  | ⟨false, i⟩, ⟨false, j⟩
+  | ⟨true, i⟩, ⟨true, j⟩ =>
+    if h : i = j then isTrue (Sigma.eq rfl h) else isFalse fun | rfl => h rfl
+
+namespace LinearMap
+
+open scoped InnerProductSpace
+
+variable {𝕜 : Type*} [RCLike 𝕜]
+
+section
+
+variable {E : Type*} [NormedAddCommGroup E] [InnerProductSpace 𝕜 E]
+
+section
+
+variable [FiniteDimensional 𝕜 E] [Fintype n] [DecidableEq n]
+
+lemma toMatrixOrthonormal_apply_apply (b : OrthonormalBasis n 𝕜 E) (f : Module.End 𝕜 E)
+    (i j : n) :
+    toMatrixOrthonormal b f i j = ⟪b i, f (b j)⟫_𝕜 :=
+  calc
+    _ = b.repr (f (b j)) i := f.toMatrix_apply ..
+    _ = ⟪b i, f (b j)⟫_𝕜 := b.repr_apply_apply ..
+
+lemma toMatrixOrthonormal_reindex [Fintype m] [DecidableEq m]
+    (b : OrthonormalBasis m 𝕜 E) (e : m ≃ n) (f : Module.End 𝕜 E) :
+    toMatrixOrthonormal (b.reindex e) f = Matrix.reindex e e (toMatrixOrthonormal b f) :=
+  Matrix.ext fun i j =>
+    calc toMatrixOrthonormal (b.reindex e) f i j
+      _ = (b.reindex e).repr (f (b.reindex e j)) i := f.toMatrix_apply ..
+      _ = b.repr (f (b (e.symm j))) (e.symm i) := by simp
+      _ = toMatrixOrthonormal b f (e.symm i) (e.symm j) := Eq.symm (f.toMatrix_apply ..)
```

Comment:
Could you make a new PR with just these two lemmas, and put them right below `def toMatrixOrthonormal`?

Also, I think both proofs can be greatly compressed; `rfl` probably works for the first one, and `Matrix.ext fun i j => rfl` for the second.